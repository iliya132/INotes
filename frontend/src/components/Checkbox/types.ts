/* eslint-disable no-unused-vars */
export interface ICheckboxProps {
    defaultValue?: boolean;
    id: string;
    className?: string;
    onChanged? : (newValue:boolean) => void;
    title?: string;
}
