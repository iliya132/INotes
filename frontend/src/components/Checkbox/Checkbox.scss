@import '../../Misc/colors.scss';

.checkbox-container {
    display: inline-block;
    width: 15px;
    height: 15px;
    background-color: $BlackCommon;
}

.share-checkbox {
    margin-right: 10px;
    cursor: pointer;
    user-select: none;
}
