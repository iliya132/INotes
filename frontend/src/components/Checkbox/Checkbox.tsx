import classNames from 'classnames';
import React, { useRef, useState } from 'react';
import Svg from '../Svg';
import { Icons } from '../Svg/types';
import styles from './Checkbox.scss';
import { ICheckboxProps } from './types';

export function Checkbox(props: ICheckboxProps) {
    const { defaultValue, id, className, onChanged, title } = props;
    const [isPressed, setPressed] = useState(defaultValue ? defaultValue : false);
    const realInputRef = useRef(null);

    const handleCheckboxClick = () => {
        (realInputRef!.current! as HTMLInputElement).value = (!isPressed).toString();
        (realInputRef!.current! as HTMLInputElement).checked = !isPressed;
        if (onChanged) {
            onChanged(!isPressed);
        }
        setPressed(!isPressed);
    };

    return (
        <div className={styles['share-checkbox']} onClick={handleCheckboxClick}>
            <div className={classNames(styles['checkbox-container'], className)} >
                <input type="checkbox" id={id} hidden ref={realInputRef} defaultChecked={defaultValue} />
                <div hidden={!isPressed}>
                    <Svg icon={Icons.Check} />
                </div>
            </div>
            <span>{title}</span>
        </div>
    );
}
