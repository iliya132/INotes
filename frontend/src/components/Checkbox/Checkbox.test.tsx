/* eslint-disable no-undef */
import React from 'react';
import { createRoot, Root } from 'react-dom/client';
import { Checkbox } from './Checkbox';
import { act } from 'react-dom/test-utils';

(globalThis as any).IS_REACT_ACT_ENVIRONMENT = true;

let root: Root | null = null;
let container: Element | null = null;
beforeEach(() => {
    act(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        root = createRoot(container);
    });
});

afterEach(() => {
    act(() => {
        root!.unmount();
        container!.remove();
        container = null;
    });
});

describe('Test Checkbox component', () => {
    it('can be rendered', () => {
        act(() => {
            root!.render(<Checkbox id="test-id" title="test" />);
        });

        let checkboxSpan = container!.querySelector('span');

        expect(checkboxSpan!.textContent).toEqual('test');
        expect(checkboxSpan).toMatchSnapshot('checkbox-snapshot');
    });

    it('can call events on click', () => {
        let isPressed = false;
        act(() => {
            root!.render(<Checkbox id="test-id" title="test" onChanged={(newValue) => (isPressed = newValue)} />);
        });
        const checkbox = container!.childNodes[0];
        act(() => {
            checkbox!.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        });

        expect(isPressed).toEqual(true);
    });

    it('setup className', () => {
        const expectedClassName = 'test-classname';
        act(() => {
            root!.render(<Checkbox id="test-id" title="test" className={expectedClassName} />);
        });
        let checkboxStyledContainer = container?.querySelector(`.${expectedClassName}`);

        expect(checkboxStyledContainer).not.toBeNull();
        expect(checkboxStyledContainer!.className).toContain(expectedClassName);
    });

    it('display check svg on preess', () => {
        act(() => {
            root!.render(<Checkbox id="test-id" title="test" />);
        });
        const checkbox = container!.childNodes[0];
        act(() => {
            checkbox!.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        });

        const checkBoxElement = container?.querySelector('div');
        expect(checkBoxElement?.innerHTML).toContain('#check');
        expect(checkBoxElement?.innerHTML).toContain('<svg');
        expect(checkBoxElement!.innerHTML).toMatchSnapshot("checkbox-pressed")
    });

    it('not pressed by default', () => {
        act(() => {
            root!.render(<Checkbox id="test-id" title="test" />);
        });
        const checkBoxElement = container?.querySelector('div');
        expect(checkBoxElement!.innerHTML).toMatchSnapshot("checkbox-unpressed")
    });
});
