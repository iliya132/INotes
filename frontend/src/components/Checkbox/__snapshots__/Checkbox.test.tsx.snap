// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Checkbox component can be rendered: checkbox-snapshot 1`] = `
<span>
  test
</span>
`;

exports[`Test Checkbox component display check svg on preess: checkbox-pressed 1`] = `"<div class=\\"checkbox-container\\"><input type=\\"checkbox\\" id=\\"test-id\\" hidden=\\"\\" value=\\"true\\"><div><svg class=\\"image tooltip\\" fill=\\"BlackCommon\\" width=\\"15px\\" height=\\"15\\"><use xlink:href=\\"test-file-stub#check\\"></use></svg></div></div><span>test</span>"`;

exports[`Test Checkbox component not pressed by default: checkbox-unpressed 1`] = `"<div class=\\"checkbox-container\\"><input type=\\"checkbox\\" id=\\"test-id\\" hidden=\\"\\"><div hidden=\\"\\"><svg class=\\"image tooltip\\" fill=\\"BlackCommon\\" width=\\"15px\\" height=\\"15\\"><use xlink:href=\\"test-file-stub#check\\"></use></svg></div></div><span>test</span>"`;
